File: SOC_Photon/src/myLibrary/iterate.cpp
/***************************************************
  A simple dynamic filter library

  Class code for embedded application.

  07-Jan-2015   Dave Gutz   Created
  30-Sep-2016   Dave Gutz   LeadLagTustin
  23-Nov-2016   Dave Gutz   LeadLagExp
 ****************************************************/
#include "iterate.h"
#include "math.h"

// class Iterator
// constructors
Iterator::Iterator() {}
Iterator::  Iterator(const String desc)
    :count_(0), desc_(desc), de_(0), des_(0), dx_(0), e_(0), ep_(0), limited_(false), x_(0), xmax_(0), xmin_(0), xp_(0)
{}
Iterator::~Iterator() {}
// operators
// functions
void Iterator::init(const double xmax, const double xmin, const double eInit)
{
    xmax_ = xmax;
    xmin_ = xmin;
    e_ = eInit;
    ep_ = eInit;
    xp_ = xmax;
    x_ = xmin;   // Do min and max first
    dx_ = x_ - xp_;
    de_ = e_ - ep_;
    count_ = 0;
}

// Generic iteration calculation, method of successive approximations for
// success_count then Newton-Rapheson as needed - works with iterateInit.
// Inputs:  e_
// Outputs:  x_
double Iterator::iterate(const boolean verbose, const uint16_t success_count, const boolean en_no_soln)
{
    de_ = e_ - ep_;
    des_  = sgn(de_)*max(abs(de_), 1e-16);
    dx_   = x_ - xp_;
    if ( verbose )
    {
        Serial.printf("%s(%d): xmin%12.8f x%12.8f xmax%12.8f e%12.8f  des%12.8f dx%12.8f de%12.8f\n", desc_.c_str(), count_, xmin_, x_, xmax_, e_, des_, dx_, de_);
    }

    // Check min max sign change
    boolean no_soln = false;
    if ( count_ == 2 )
    {
        if ( e_*ep_ >= 0  && en_no_soln ) // No solution possible
        {
            no_soln = true;
            if ( abs(ep_) < abs(e_) )
                x_   = xp_;
            ep_  = e_;
            limited_ = false;
            if ( verbose )
                Serial.printf("%s:No soln\n", desc_.c_str());  // Leaving x at most likely limit value and recalculating...
            return ( e_ );
        }
        else
            no_soln = false;
    }
    if ( count_==3 && no_soln )  // Stop after recalc and no_soln
    {
        e_ = 0;
        return ( e_ );
    }
    xp_ = x_;
    ep_ = e_;
    if ( count_ == 1 )
        x_ = xmax_;  // Do min and max first
    else
    {
        if ( count_ > success_count )
        {
            x_ = max(min(x_ - e_/des_*dx_, xmax_),xmin_);
            if ( e_ > 0 )
                xmax_ = xp_;
            else
                xmin_ = xp_;
        }
        else
        {
            if ( e_ > 0 )
            {
                xmax_ = xp_;
                x_ = (xmin_ + x_)/2;
            }
            else
            {
                xmin_ = xp_;
                x_ = (xmax_ + x_)/2;
            }
        }
        if ( x_==xmax_ || x_==xmin_ )
            limited_ = 0;
        else
            limited_ = 1;
    }
    return ( e_ );
}

File: SOC_Photon/py/Iterate.py
# Iterate class to provide a general purpose solver.   Begins with successive approximation and finishes with Newton-Rapheson
# Copyright (C) 2023 Dave Gutz
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# See http://www.fsf.org/licensing/licenses/lgpl.txt for full license text.

__author__ = 'Dave Gutz <davegutz@alum.mit.edu>'
__version__ = '$Revision: 1.1 $'
__date__ = '$Date: 2023/03/02 13:15:02 $'

import numpy as np


class Iterator:
    def __init__(self, desc=''):
        self.count = 0
        self.desc = desc
        self.de = 0.
        self.des = 0.
        self.dx = 0.
        self.e = 0.
        self.ep = 0.
        self.limited = False
        self.x = 0.
        self.xmax = 0.
        self.xmin = 0.
        self.xp = 0.

    # Initialize
    def init(self, xmax=0., xmin=0., eInit=0.):
        self.xmax = xmax
        self.xmin = xmin
        self.e = eInit
        self.ep = self.e
        self.xp = self.xmax
        self.x = self.xmin  # Do min and max first
        self.dx = self.x - self.xp
        self.de = self.e - self.ep
        self.count = 0

    # Generic iteration calculation, method of successive approximations for
    # success_count then Newton-Rapheson as needed - works with iterateInit.
    # Inputs:  self.e
    # Outputs:  self.x
    def iterate(self, verbose=False, success_count=0, en_no_soln=False):
        self.de = self.e - self.ep
        self.des = np.sign(self.de)*max(abs(self.de), 1e-16)
        self.dx = self.x - self.xp
        if verbose:
            s = self.desc + '(' + "{:2d}".format(self.count) + ':'
            s += " xmin{:12.8f}".format(self.xmin)
            s += " xmax{:12.8f}".format(self.xmax)
            s += " e{:12.8f}".format(self.e)
            s += " des{:12.8f}".format(self.des)
            s += " dx{:12.8f}".format(self.dx)
            s += " de{:12.8f}".format(self.de)
            s += " e{:12.8f}".format(self.e)
            # print(s)

        # Check min max sign change
        no_soln = False
        if self.count == 2:
            if self.e*self.ep >= 0 and en_no_soln:  # No solution possible
                no_soln = True
            if abs(self.ep) < abs(self.e):
                self.x = self.xp
            self.ep = self.e
            self.limited = False
            if verbose:
                print(self.desc, ':No soln')  # Leaving x at most likely limit value and recalculating...
            return self.e
        else:
            no_soln = False

        # Stop after recalc and no_soln
        if self.count == 3 and no_soln is True:
            self.e = 0.
            return self.e
        self.xp = self.x
        self.ep = self.e
        if self.count == 1:
            self.x = self.xmax  # Do min and max first
        else:
            if self.count > success_count:
                self.x = max(min(self.x - self.e/self.des*self.dx, self.xmax), self.xmin)
                if self.e > 0:
                    self.xmax = self.xp
                else:
                    self.xmin = self.xp
            else:
                if self.e > 0:
                    self.xmax = self.xp
                    self.x = (self.xmin + self.x) / 2.
                else:
                    self.xmin = self.xp
                    self.x = (self.xmax + self.x) / 2.
            if self.x == self.xmax or self.x == self.xmin:
                self.limited = False
            else:
                self.limited = True
        return self.e
